crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs

```diff
@@ -123,6 +125,80 @@ impl<'a> FunctionContext<'a> {
             ast::Type::Vec(_) => Type::Reference,
         }
     }
+
+    /// Insert a unit constant into the current function if not already
+    /// present, and return its value
+    pub(super) fn unit_value(&mut self) -> Values {
+        self.builder.numeric_constant(0u128.into(), Type::Unit).into()
+    }
+
+    /// Insert a binary instruction at the end of the current block.
+    /// Converts the form of the binary instruction as necessary
+    /// (e.g. swapping arguments, inserting a not) to represent it in the IR.
+    /// For example, (a <= b) is represented as !(b < a)
+    pub(super) fn insert_binary(
+        &mut self,
+        mut lhs: ValueId,
+        operator: noirc_frontend::BinaryOpKind,
+        mut rhs: ValueId,
+    ) -> Values {
+        let op = convert_operator(operator);
+
+        if operator_requires_swapped_operands(operator) {
+            std::mem::swap(&mut lhs, &mut rhs);
+        }
+
+        // TODO: Rework how types are stored.
+        // They should be on values rather than on instruction results
+        let typ = Type::field();
+        let mut result = self.builder.insert_binary(lhs, op, rhs, typ);
+
+        if operator_requires_not(operator) {
+            result = self.builder.insert_not(result, typ);
+        }
+        result.into()
+    }
+}
+
+/// True if the given operator cannot be encoded directly and needs
+/// to be represented as !(some other operator)
+fn operator_requires_not(op: noirc_frontend::BinaryOpKind) -> bool {
+    use noirc_frontend::BinaryOpKind::*;
+    matches!(op, NotEqual | LessEqual | GreaterEqual)
+}
+
+/// True if the given operator cannot be encoded directly and needs
+/// to have its lhs and rhs swapped to be represented with another operator.
+/// Example: (a > b) needs to be represented as (b < a)
+fn operator_requires_swapped_operands(op: noirc_frontend::BinaryOpKind) -> bool {
+    use noirc_frontend::BinaryOpKind::*;
+    matches!(op, Greater | LessEqual)
+}
+
+/// Converts the given operator to the appropriate BinaryOp.
+/// Take care when using this to insert a binary instruction: this requires
+/// checking operator_requires_not and operator_requires_swapped_operands
+/// to represent the full operation correctly.
+fn convert_operator(op: noirc_frontend::BinaryOpKind) -> BinaryOp {
```

Comment:
Initially I wrote: Can this function return `(BinaryOp, Option<BinaryOp>) ` to signify the fact that some frontend operators, need two ssa operators to represent it. 

i.e. 
```rust
BinaryOpKind::Add => (BinaryOp::Add, None),
BinaryOpKind::NotEqual => (BinaryOp::Eq, Some(BinaryOp::Not)),
```
 
 With the motivation of not needing `operator_requires_not` and `operator_requires_swapped_operands` incase one forgets to call those functions. But actually it seems that we will need it due to that not being enough to represent the fact that some operations will require one to swap operands.
 
I am okay with it because I think we don't need to call this in a bunch of places; only when converting a monomorphized ast binary op expression to an ssa binary op expression

Reply:
Yes, this should only be needed in this one case. Originally I had a large match table of each operator, its lhs, rhs, and whether it requires a not:
```rs
let (lhs, op, rhs, requires_not) = match operator {
    Add => (lhs, BinaryOp::Add, rhs, false),
    Sub => (lhs, BinaryOp::Sub, rhs, false),
    ...
    Gt => (rhs, BinaryOp::Lt, lhs, false),
    Gte => (lhs, BinaryOp::Lt, rhs, true),
    Lte => (rhs, BinaryOp::Lt, lhs, true),
    ...
}
```
But I decided to split it up into 3 functions instead